### PR TITLE
Add audit log

### PR DESCRIPTION
**Pull request by michaelschattgen**
This pull request is aiming to give more insight to important events that occurred within the app. We chose to use a SQLite database in combination with Room. I've made the choice to only show the entries of the last 30 days for now, but we might want to change that in the future.

The events we're currently tracking are:
- Vault unlocked
- Backup created
- Vault exported
- Entry shared
- Failed vault unlock attempt with password
- Failed vault unlock attempt with biometrics

Feel free to comment on this PR if you think we need to track other events.

This is discussed in #1069 and makes it possible to implement #1132.

### CODE REVIEW

**Comment by InfiniteCoder06** [comment]
@michaelschattgen A small suggestion is to delete the entries greater than 30 days periodically.

File: app/src/main/java/com/beemdevelopment/aegis/ui/views/AuditLogAdapter.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+        int position = getItemCount() - 1;
+        if (position == 0) {
+            notifyDataSetChanged();
+        } else {
+            notifyItemInserted(position);
+        }
+    }
+
+    public void addReferencedEntry(VaultEntry vaultEntry) {
```
Comment (alexbakker) [inline app/src/main/java/com/beemdevelopment/aegis/ui/views/AuditLogAdapter.java:40]: Do we need to separately add the referenced vault entries here? Perhaps we can introduce a ``AuditLogEntryModel`` type that holds the ``AuditLogEntry`` and an optional referenced ``VaultEntry``.

File: app/src/main/java/com/beemdevelopment/aegis/ui/views/AuditLogHolder.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    private final CardView _cardView;
+    private final ImageView _auditLogEntryIcon;
+    private final TextView _auditLogEntryTitle;
+    private final TextView _auditLogEntryDescription;
+    private final TextView _auditLogEntryReference;
+    private final TextView _auditLogEntryTimestamp;
+    private final int errorBackgroundColor;
+    private final ColorStateList initialBackgroundColor;
+    private final int initialIconColor;
```
Comment (alexbakker) [inline app/src/main/java/com/beemdevelopment/aegis/ui/views/AuditLogHolder.java:34]: Style nit.

File: app/src/main/java/com/beemdevelopment/aegis/ui/AuthActivity.java [hunk 3]
```diff
@@ -354,6 +356,7 @@ private class BiometricPromptListener extends BiometricPrompt.AuthenticationCall
         public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
             super.onAuthenticationError(errorCode, errString);
             _bioPrompt = null;
+            _auditLogManager.addVaultUnlockFailedPasswordEvent();
```
Comment (alexbakker) [inline app/src/main/java/com/beemdevelopment/aegis/ui/AuthActivity.java:359]: You probably meant to call ``addVaultUnlockFailedBiometricsEvent()`` here. We should probably only log the event if the biometrics prompt was not cancelled by the user.

**Review by alexbakker — CHANGES_REQUESTED** [review]
This looks great! I've added a couple of comments.

I think the audit log event types you defined cover most of the important events, but I am missing an event for backups made through ``AegisBackupAgent``. Because Android starts backup agents in a weird restricted mode, this one requires some careful testing to make sure it works correctly.

**Comment by elliotwutingfeng** [comment]
I like this feature! I think it should also track when the vault password has been changed.

**Comment by michaelschattgen** [comment]
@elliotwutingfeng I like that! I'm leaving this pull request for what it is (otherwise it will never get done) but that's definitely going on the backlog. Thanks for the suggestion.

**Review by alexbakker — APPROVED** [review]
Very nice!

**Pull request merged**